lib/Sema/CSBindings.cpp

```diff
@@ -400,6 +400,11 @@ ConstraintSystem::getPotentialBindings(TypeVariableType *typeVar) {
     if (type->hasError())
       continue;
 
+    // If we're not allowed to bind to an InOut type, try to bind to the
+    // underlying type instead.
+    if (!typeVar->getImpl().canBindToInOut() && typeVar->getImpl().canBindToLValue())
```

Comment:
This seems wrong...

Reply:
Oddly enough, it does the right thing.  For this, the alternatives is to let the binding through: Then we bind `x` to `inout Int` transitively and `matchTypes` never has a chance to suggest the fix.

```swift
func f0(_ x: inout Int) {}
let g = { x in f0(x) }
```

Reply:
Which is a symptom of `inout` inference being a thing at all with closures.  It's not valid in general to propagate from the declref back to the parameter if there's an intervening `inout`.  We should instead be trying to bind the parameter from the closure to the call but that's a larger change.

Reply:
Closures already use separate type variables for the 'external' parameter type and 'internal' parameter type. For example, if the 'external' type is inout T, the 'internal' type is @lvalue T. They are related with a BindParam constraint.

Maybe this is not plumbed all the way through. Can you show me a test case?

Reply:
Closure inference specifically does not open the type of the declref for closure parameters and throws in its own `BindParam` constraint.

Reply:
That sounds wrong... I wonder if it causes problems elsewhere. Can you point me at the code?

Reply:
Conveniently, this diff includes the salient part of `ConstraintSystem.cpp`

Reply:
That's just the part where we create a second 'internal' type variable and relate it with the original 'external' type variable via a BindParam constraint, which is what I was talking about above.

Reply:
so, can you figure out what's going wrong here? :-) I'm not convinced your change is the correct fix for the problem.

Reply:
Well that constraint is the root of all this trouble.  I'll revert the last commit and come up with something more clever

Reply:
Could it be that 'edge contraction' or some other optimization pass is performing an invalid transformation on the constraint graph before the solver runs?

Reply:
I thought @xedin stopped that kind of contraction from being performed a while ago.